### PR TITLE
fix(gatsby): Catch errors when persisting state

### PR DESCRIPTION
**Pull request by stefanprobst**
When persisting state, await the `saveState` promises and catch potential errors.

### CODE REVIEW

File: packages/gatsby/src/db/index.js [hunk 1]
```diff
@@ -10,10 +11,18 @@ if (process.env.GATSBY_DB_NODES === `loki`) {
 }
 
 // calls `saveState()` on all DBs
-function saveState() {
-  for (const db of dbs) {
-    db.saveState()
+let saveInProgess = false
```
Comment (sidharthachatterjee) [inline packages/gatsby/src/db/index.js]:
```suggestion
let saveInProgress = false
```

File: packages/gatsby/src/db/index.js [hunk 2]
```diff
@@ -10,10 +11,18 @@ if (process.env.GATSBY_DB_NODES === `loki`) {
...
+async function saveState() {
+  if (saveInProgess) return
```
Comment (sidharthachatterjee) [inline packages/gatsby/src/db/index.js]:
```suggestion
  if (saveInProgress) return
```

File: packages/gatsby/src/db/index.js [hunk 3]
```diff
@@ -10,10 +11,18 @@ if (process.env.GATSBY_DB_NODES === `loki`) {
...
+  saveInProgess = true
```
Comment (sidharthachatterjee) [inline packages/gatsby/src/db/index.js]:
```suggestion
  saveInProgress = true
```

File: packages/gatsby/src/db/index.js [hunk 4]
```diff
@@ -10,10 +11,18 @@ if (process.env.GATSBY_DB_NODES === `loki`) {
...
+
+  try {
+    await Promise.all(dbs.map(db => db.saveState()))
+  } catch (err) {
+    report.warn(`Error persisting state: ${(err && err.message) || err}`)
   }
+
+  saveInProgess = false
```
Comment (sidharthachatterjee) [inline packages/gatsby/src/db/index.js]:
```suggestion
  saveInProgress = false
```

**Comment by stefanprobst** [comment]
Thanks @sidharthachatterjee! Fixed.

**Comment by sidharthachatterjee** [comment]
Published in gatsby@2.1.18

**Pull request merged**